src/trel_dnssd/trel_dnssd.cpp

```diff
@@ -0,0 +1,542 @@
+
+    otPlatTrelHandleDiscoveredPeerInfo(mNcp.GetInstance(), &peerInfo);
+}
+
+void TrelDnssd::RemoveAllPeers(void)
+{
+    for (const auto &entry : mPeers)
+    {
+        NotifyRemovePeer(entry.second);
+    }
+
+    mPeers.clear();
+}
+
+void TrelDnssd::CheckTrelNetifReady(void)
+{
+    assert(IsInitialized());
+
+    if (mTrelNetifIndex == 0)
+    {
+        mTrelNetifIndex = if_nametoindex(mTrelNetif.c_str());
+
+        if (mTrelNetifIndex != 0)
+        {
+            otbrLogDebug("Netif %s is ready: index = %d", mTrelNetif.c_str(), mTrelNetifIndex);
+            OnBecomeReady();
+        }
+        else
+        {
+            uint16_t delay = kCheckNetifReadyIntervalMs;
+
+            otbrLogWarning("Netif %s is not ready, will retry after %d seconds", mTrelNetif.c_str(), delay / 1000);
+            mTaskRunner.Post(Milliseconds(delay), [this]() { CheckTrelNetifReady(); });
+        }
+    }
+}
+
+bool TrelDnssd::IsReady(void) const
+{
+    assert(IsInitialized());
+
+    return mTrelNetifIndex > 0 && mMdnsPublisherReady;
+}
+
+void TrelDnssd::OnBecomeReady(void)
+{
+    if (IsReady())
+    {
+        otbrLogInfo("TREL DNS-SD Is Now Ready: Netif=%s(%u), SubscriberId=%u, Register=%s!", mTrelNetif.c_str(),
+                    mTrelNetifIndex, mSubscriberId, mRegisterInfo.mInstanceName.c_str());
+
+        if (mSubscriberId > 0)
+        {
+            mPublisher.SubscribeService(kTrelServiceName, "");
+        }
+
+        if (mRegisterInfo.IsValid())
+        {
+            DoRegisterTrelService();
+        }
+    }
+}
+
+uint16_t TrelDnssd::CountDuplicatePeers(const TrelDnssd::Peer &aPeer)
+{
+    uint16_t count = 0;
+
+    for (const auto &entry : mPeers)
+    {
+        if (&entry.second == &aPeer)
+        {
+            continue;
+        }
+
+        if (!memcmp(&entry.second.mSockAddr, &aPeer.mSockAddr, sizeof(otSockAddr)) &&
+            !memcmp(&entry.second.mExtAddr, &aPeer.mExtAddr, sizeof(otExtAddress)))
+        {
+            count++;
+        }
+    }
+
+    return count;
+}
+
+void TrelDnssd::RegisterInfo::Assign(uint16_t aPort, const uint8_t *aTxtData, uint8_t aTxtLength)
+{
+    assert(!IsRegistered());
+    assert(aPort > 0);
+
+    mPort = aPort;
+    mTxtEntries.clear();
+
+    for (uint8_t r = 0; r < aTxtLength;)
+    {
+        uint8_t entrySize = aTxtData[r];
+        uint8_t keyStart  = r + 1;
+        uint8_t entryEnd  = keyStart + entrySize;
+        uint8_t keyEnd    = keyStart;
+        uint8_t valStart;
+
+        while (keyEnd < entryEnd && aTxtData[keyEnd] != '=')
+        {
+            keyEnd++;
+        }
+
+        valStart = keyEnd;
+        if (valStart < entryEnd && aTxtData[valStart] == '=')
+        {
+            valStart++;
+        }
+
+        mTxtEntries.emplace_back(reinterpret_cast<const char *>(&aTxtData[keyStart]), keyEnd - keyStart,
+                                 &aTxtData[valStart], entryEnd - valStart);
+
+        r += entrySize + 1;
+        assert(r <= aTxtLength);
+    }
```

Comment:
It's fine to keep it as is. But in the future, could you extract this snippet and make it a util function?

Reply:
Mdns::Publisher::DecodeTxtData implemented.